Create 2020-04-22-two-phased-canary-rollout-with-gloo.md

rickducott opened the pull request:
This is a post detailing an operational workflow, a two-phased approach to correctness and load testing a new version of an application, behind open-source Gloo as an API gateway for Kubernetes.

I tried to follow some existing posts but please let me know how I can improve this before publication!

rickducott commented:
/assign @zacharysarah

zacharysarah commented:
/unassign
/assign @kbarnard10

rickducott commented:
@kbarnard10 any update or eta here?

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+---
+title: " Two-phased Canary Rollout with Open Source Gloo "
```
nit:
```suggestion
title: "Two-phased Canary Rollout with Open Source Gloo"
```

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+slug: two-phased-canary-rollout-with-gloo
+url: /blog/2020/03/Two-phased-Canary-Rollout-With-Gloo
+---
+
```
On this blog, posts usually show the author name and optionally their organization.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+ version `v1` of an example application to the `echo` namespace. We'll expose this application outside the cluster 
+ by creating a route in Gloo, to end up with a picture like this: 
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/1-setup/setup.png)
```
1. This image should have an alt attribute.
2. Can you copy the image into the repo so that it's served via the content CDN and the image is in source control?

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+Now, you should be able to run `glooctl version` to see that it is installed correctly:
+
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+The console should indicate the install finishes successfully:
+
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+Before long, we can see all the Gloo pods running in the `gloo-system` namespace:
+
+```bash
````
This isn't bash code (the first line is shell plus a prompt though)

rickducott replied:
thanks - i went and just made these all plain code blocks. open to other ideas for presenting the command + output, I went with this one and tried to use it everywhere

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+            - containerPort: 8080
+```
+
+And here is the `echo` Kubernetes `Service` object:
````
nit:
```suggestion
And here is the `echo` Service, a Kubernetes object:
```
?

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+    app: echo
+```
+
+For convenience, we've published this yaml in a repo so we can deploy it with the following command:
````
I suggest:
```suggestion
For convenience, we've published this manifest so you can deploy it with the following command:
```

and also that you put it in the website repo. See https://kubernetes.io/docs/contribute/style/write-new-topic/#including-code-from-another-file

(you can use the website repo to store the manifest even if you decide not to use a `codenew` shortcode).

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+```
+
+We should see the following output:
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+```
+
+And we should be able to see all the resources healthy in the `echo` namespace:
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+Once we apply these two resources, we can start to send traffic to the application through Gloo:
+
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+traffic to the new version. We'll use this header to perform some basic smoke testing and make sure `v2` is working the 
+way we'd expect:
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/2-initial-subset-routing-to-v2/subset-routing.png)
```
1. This image should have an alt attribute.
2. Can you copy the image into the repo so that it's served via the content CDN and the image is in source control?

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+The application should continue to function as before:
+
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+Make sure `v2` is running before moving on:
+
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+The application should continue to function as before:
+
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+We'll route to the `v2` subset when the `stage: canary` header is supplied on the request. If the header isn't 
+provided, we'll continue to route to the `v1` subset as before.  
+
+```yaml
+apiVersion: gateway.solo.io/v1
+kind: VirtualService
+metadata:
+  name: echo
+  namespace: gloo-system
+spec:
+  virtualHost:
+    domains:
+      - '*'
+    routes:
+      - matchers:
+          - headers:
+              - name: stage
+                value: canary
+            prefix: /
+        routeAction:
+          single:
+            upstream:
+              name: echo
+              namespace: gloo-system
+            subset:
+              values:
+                version: v2
+      - matchers:
+          - prefix: /
+        routeAction:
+          single:
+            upstream:
+              name: echo
+              namespace: gloo-system
+            subset:
+              values:
+                version: v1
+```
````
I recommend copying this manifest into the repo, as per an earlier comment.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+Now that we have this route, we can do some testing. First let's ensure that the existing route is working as expected:
+
+```bash
````
This isn't bash code

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+And now we can start to canary test our new application version:
+
+```bash
````
This isn't bash code

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+With the initial weights, we should see the gateway continue to serve `v1` for all traffic.
+
+```bash
````
This isn't bash code

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
````diff
@@ -0,0 +1,772 @@
+
+Now when we send traffic to the gateway, we should see all of the requests return `version:v2`. 
+
+```bash
````
This isn't bash code.

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+
+Now our cluster looks like this:
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/4-decommissioning-v1/end-state.png)
```
1. This image should have an alt attribute.
2. Can you copy the image into the repo so that it's served via the content CDN and the image is in source control?

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+Running through these steps over a 15 minute period while sending a significant load to these routes on the proxy 
+yielded the following operational dashboards:
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/perf-test.png)
```
1. This image should have an alt attribute.
2. Can you copy the image into the repo so that it's served via the content CDN and the image is in source control?

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+
+We will continue adjusting weights until eventually, all of the traffic is now being routed to `v2`:
+
+![](https://raw.githubusercontent.com/solo-io/gloo-ref-arch/blog-30-mar-20/platform/prog-delivery/two-phased-with-os-gloo/3-progressive-traffic-shift-to-v2/final-shift.png)
```
1. This image should have an alt attribute.
2. Can you copy the image into the repo so that it's served via the content CDN and the image is in source control?

sftim commented on content/en/blog/_posts/2020-03-30-two-phased-canary-rollout-with-gloo.md:
```diff
@@ -0,0 +1,772 @@
+* Check out the [docs](https://docs.solo.io/gloo/latest), which have an extensive collection of guides and examples
+* Join the [slack channel](http://slack.solo.io/) and start chatting with the Solo engineering team and user community
+
+If you'd like to get in touch with me (feedback is always appreciated!), you can find me on Slack or email me at 
+**rick.ducott@solo.io**. 
```
Which Slack? Solo Slack or Kubernetes Slack? Or both?

rickducott replied:
Thanks - technically either but I linked to solo slack :D

sftim commented:
@rickducott did you have a chance to look at the feedback on this PR? 

@castrojo / @parispittman have you got any other feedback for Rick?

castrojo commented:
The over-bolding on the intro and outro doesn't match most of our style guidelines so I would strip those out. Mostly looks good, please address this and the reviewer concerns and I'd be happy to lgtm.

rickducott commented:
Thanks for taking a look! I am going to update based on the feedback today.

rickducott commented:
@sftim @castrojo thanks so much for the feedback. I made all the updates requested -- added author, moved images to repo, updated code blocks, added alt texts, removed excessive intro bolding -- and fixed the date. LMK if you'd like any other changes!

castrojo commented:
/lgtm

We just need to sort the final date in the filename/slug to be an upcoming date.

castrojo commented:
/lgtm

Thanks for making these adjustments!
